runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java

```diff
@@ -93,14 +110,38 @@ public Layout getCurrentLayout() {
                     log.warn("Got a wrong epoch exception, updating epoch to {} and "
                             + "invalidate view", we.getCorrectEpoch());
                     runtime.invalidateLayout();
+                } else if (re instanceof NetworkException) {
+                    log.warn("layoutHelper: System seems unavailable", re);
+
+                    runtime.systemDownHandler.run();
+                    runtime.invalidateLayout();
+
+                    try {
+                        Thread.sleep(runtime.retryRate * 1000);
+                    } catch (InterruptedException e) {
+                        log.warn("Interrupted Exception in layout helper.", e);
+                    }
+
                 } else {
                     throw re;
                 }
+                if (rethrowAllExceptions) {
+                    throw new RuntimeException(re);
+                }
+
             } catch (InterruptedException ie) {
                 throw new UnrecoverableCorfuInterruptedError("Interrupted in layoutHelper", ie);
             } catch (ExecutionException ex) {
                 log.warn("Error executing remote call, invalidating view and retrying in {}s",
                         runtime.retryRate, ex);
+
+                // If SystemUnavailable exception is thrown by the layout.get() completable future,
+                // the exception will materialize as an ExecutionException. In that case, we need to propagate
+                // this exception.
+                if (ex.getCause() instanceof SystemUnavailableException) {
+                    throw (SystemUnavailableException) ex.getCause();
+                }
+
                 runtime.invalidateLayout();
                 Utils.sleepUninterruptibly(runtime.retryRate * 1000);
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Cast one of the operands of this multiplication operation to a "long". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2184)